File: 2019/B/subarray.cc
#include <bits/stdc++.h>

using namespace std;

const int N_max = 1e5;
int T;
int N;
int S;
int A[N_max];
int counts[N_max];
int scores[N_max];

void init_counts() {
  for(int i = 0; i < N_max; i++) {
    counts[i] = 0;
  }
}

void comp_scores(int l) {
  int a, c;
  for(int i=0; i < N - l; i++){
    a = A[i + l];
    c = counts[a];
    counts[a] += 1;
    if(c < S) {
      scores[i] = 1;
    }
    if(c > S) {
      scores[i] = 0;
    }
    if(c == S) {
      scores[i] = -S;
    }
  }
}

int max_score(int l) {
  init_counts();
  comp_scores(l);

  int e = 0;
  int sum = 0;
  for(int i = 0; i < N - l; i++) {
    sum += scores[i];
    if(sum > e) {
      e = sum;
    }
  }
  return e;
}

int main() {
  scanf("%d\n", &T);
  for(int t = 1; t <= T; t++) {
    scanf("%d%d", &N, &S);
    for(int n = 0; n < N; n++) {
      scanf("%d", &A[n]);
    }
    int e = 0;
    for(int l = 0; l < N; l++) {
      e = max(max_score(l), e);
    }
    printf("Case #%d: %d\n", t, e);
  }
}

File: 2019/B/subarray.py
import numpy as np

#
# FUNCTIONS
#

def scores(l):
    counts = {}
    scores = np.empty(N - l, dtype=int)
    for i in range(N - l):
        a = A[i + l]
        c = counts.setdefault(a, 0)
        counts[a] = c + 1
        if c < S:
            scores[i] = 1
        if c > S:
            scores[i] = 0
        if c == S:
            scores[i] = -S
    return scores

def max_score(s):
    e = 0
    sum = 0
    for i in range(len(s)):
        sum += s[i]
        if sum > e:
            e = sum
    return e

#
# INPUT
#

# Read number of test cases
T = int(input())

for t in range(1, T + 1):
    # Read variables
    ################

    N, S = map(int, input().split())
    A = np.fromstring(input(), dtype=int, sep=' ')

    # Calc output
    #############
    e = 0
    for l in range(N):
        s = scores(l)
        e = max(max_score(s), e)

    # Print result
    ##############
    print("Case #{}: {}".format(t, e))
